### ISSUE TITLE
AWS Lambda Language

### ISSUE THREAD

**Issue by craiglonsdale** [breadtubetv/breadtubetv#59]
As we start to move to some more advancd functionality, we should probably make a decision on what language we want our serverless stack code to be run on.
Netify uses AWS Lambdas and they support
- C#
- NodeJS
- Python
- Java
- Go
- Ruby
- Powershell

I think that C# is the lowest barrier of entry for new devs, but that doesn't mean it's a good choice. I'm happy to pick up a new language if necessary.

What are some thoughts on this?

**Comment by dirkkelly** [comment]
I would say Ruby or Python would be the lowest barrier of entry.

I think it should be Go so that we have project parity.

Voted.

**Comment by j16r** [comment]
I vote Go - it's the easiest to learn of all of them, maintains project parity and isn't tied to a mostly closed platform (Microsoft).

**Comment by soliveira-vouga** [comment]
I'd argue NodeJs is the easiest to learn out of all of them since most developers already know javascript, but I lean more toward using Go simply because I've never used it and wouldn't mind learning it

**Comment by j16r** [comment]
I will teach you all!

**Comment by craiglonsdale** [comment]
Looks like it's Go, as per votes by active devs.
I for one welcome our new programming language overlord.

**Comment by soliveira-vouga** [comment]
I think the cool thing about FaaS such as lambda and the concept of microservices is that your microservice or function can use whatever language you want. If @craiglonsdale decides to start a new microservice using Go, that service uses Go. If @dirkkelly creates another microservice using C# then that's cool too. They all work with Lambda. So I'll change my vote to let's remain language agnostic

**Comment by j16r** [comment]
@soliveira-vouga true, though it would mean that you can't share any code and any newcomers would potentially have to learn N languages, rather than just the one.

**Comment by j16r** [comment]
@soliveira-vouga true, though it would mean that you can't share any code and any newcomers would potentially have to learn N languages, rather than just the one.

**Comment by dirkkelly** [comment]
I would love to see us deploy https://github.com/breadtubetv/breadtubetv/pull/51 so that we have a basis to work from, for a couple of reasons:

- There is simple changes needed, which will help you learn the framework/language
- We can automate channels after this merge
- We can use that automated channel to learn how forms and lambdas work together
- Adding any additional script (regardless of language) will leverage this add channel lambda

@soliveira-vouga, @j16r, @craiglonsdale are any of you able to spend some time this week on it? Maybe this is an opportunity to pair.

**Comment by quietlemon** [comment]
Sorry to interrupt a rather focused discussion, but why would you rely on lambda functions and AWS in the first place? FaaS in principle is not really a power-structure problem (as you can actually self-host OpenFaaS and others) but it's far from easy/optimal/ecological. The kind of setup you need to get FaaS running and properly contained adds a lot of layers of complexity to what could be a simple script (hence [throwing away resources](http://tonsky.me/blog/disenchantment/)), and then running functions on the network is just consuming network bandwidth (and electricity) for dubious reasons.

Now simpler (and arguably more ecological and anarchist) approaches to what you're looking for could be:

- webhooks (on your forge) to trigger a CI build when needed?
- cron tasks?
- a server-side webapp opening PRs on your forge?
- adding steps in your build/deploy pipeline?
- a mix of all this?

I'm guessing this decision was partly dictated by the host being Netlify. But probably this issue should be addressed as well, in favor of non-profit hosting coops or self-hosting.

So in the end it depends on what you're trying to build, but if it's just about the content importing pipelines, in the end there's much easier, resilient, and ecological ways to achieve your goals. FaaS is **in most cases** barely more than a capitalist trend to centralize computing (and power) in the hands of a few.

I don't mean to come up as rude. I hope my criticism was constructive enough. If not, feel free to ask questions or criticize my patronizing :)

I wish you all the best! Keep the revolutionary struggle going throughout cyberspace!

**Comment by dirkkelly** [comment]
@quietlemon hey there. Not sure if you're still using your account but would love to discuss this and all your other great posts with you more. If you are around can you hit me up on https://breadtube.tv/discord

We now have working go import scripts and have started to get quite a few submissions at a time, designing an effective and sustainable ecosystem for automating this process would be really helpful for making the site more reactive and useful to the community.
